wireapp/wire-android
# feat: add anonymous analytics (WPB-8933)

<table>
<td>
  <a href="https://wearezeta.atlassian.net/browse/WPB-8933" title="WPB-8933" target="_blank"><img alt="Task" src="https://wearezeta.atlassian.net/rest/api/2/universal_avatar/view/type/issuetype/avatar/10818?size=medium" />WPB-8933</a>  [Android] Integrate countly SDK
  </td></table>
  <br />
 






----
#### PR Submission Checklist for internal contributors

- The **PR Title**

- The **PR Description**
----

# What's new in this PR?

### Issues

No anonymous analytics implemented.

### Solutions

Implementation of Anonymous Analytics [Countly]

NOTE: Most files are either: stubs, new files from new modules (.gitignore, AndroidManifest, string.xml), so don't be afraid :)

We have added 3 new modules in this PR:
`analytics`
`analytics-disabled`
`analytics-enabled`

- `analytics`
	Holds the Interfaces + Stubs and models:
	- Models {
		 - `AnalyticsEvent` -> interface where new events will be based of
		 - `AnalyticsSettings` -> data class to hold settings when initializing an analytics SDK
	}

- `analytics-disabled`
	Holds the implementation of Stubs from `analytics`, just a wrapper in case no analytics is set for the build

- `analytics-enabled`
	The real implementation of `AnalyticsManager` and `AnalyticsRecorder`
	- Recorder -> Where we really configure/initialize the analytics SDK (in this case, Countly)
	- Manager -> Wrapper + verifying if the analytics SDK can be initialized or events can be sent

- default.json
    - Added 3 new attributes to all flavors : [analytics_enabled, analytics_app_key, analytics_server_url]

- On previous PR [feat: add toggle to opt-out analytics (WPB-8922)](https://github.com/wireapp/wire-android/pull/3008) we added the toggle value on `GlobalDataStore` which for now its ok, but we need to check in the future if we should move it to User prefs.

- On WireApplication we are initializing the AnalyticsManager after our current run of LoggingFrameworks
- On WireActivity we added a lifecycle observer to fire the events on `onStart()` and `onStop()`

More importantly:
- app:build.gradle.kts -> Added a function to retrieve configs from all existing flavors and add the expected analytics module to it, example: ``` If `beta` flavor has `analytics_enabled` on its config, we will add `betaImplementation(project(":core:analytics-enabled")` otherwise it will be added the `analytics-disabled` ```

- On `AndroidLibraryConventionPlugin` are now replacing `-` to `_` on namespaces as it doesn't allow `-` but our modules contain `-` for better readability.

Needs releases with:

- [X] #3008 

### Testing

#### How to Test

- Open App with a build that enables analytics (`beta` Debug for example)
- Verify on logcat that the `app.open` event has been sent /or
- Verify on Countly that the `app.open` event has been sent

### Notes (Optional)

This is the first PR, only for integrating Countly SDK + Anonymous Analytics as a general purpose in case we want to do any changes of SDK later.
